### DISCUSSION TITLE
Cannot load OQS provider inside code

### DISCUSSION THREAD

**Discussion by IProtex** [open-quantum-safe/oqs-provider#59] [Q&A] [answered]
Hi, I followed the [Building and testing](https://github.com/open-quantum-safe/oqs-provider#building-and-testing) section within the README to create the `oqsprovider.so` module library.

The loading of the module inside my code using the following code snippet, seems to work as it does not throw an exception and the pointer is not `NULL`.
```
void *oqsprovider = dlopen("../modules/oqsprovider.so", RTLD_NOW);
```

However, if I try to load the `oqsprovider` via OpenSSL3 afterwards, it does not work as the `OSSL_PROVIDER` pointer always gets initialise with `NULL`:
```
OSSL_PROVIDER *oqs = OSSL_PROVIDER_load(NULL, "oqsprovider");
```

Any help or advice would be greatly appreciated :)

**Comment by baentsch** [comment]
Hmm -- "interesting".... Does anything change if you pass  a real libctx (`OSSL_LIB_CTX_new()`) instead of NULL? If so I wonder whether this behaviour is OK within the API spec...

**Reply by baentsch** [reply]
Just tested: Works fine for me even without libctx param: Sample code (from the doc):
```
#include <openssl/params.h>
#include <openssl/core_names.h>
#include <openssl/provider.h>
#include <openssl/err.h>

int main(int argc, char* argv[]) {
 OSSL_PROVIDER *prov = NULL;
 const char *build = NULL;
 size_t built_l = 0;
 OSSL_PARAM request[] = {
     { OSSL_PROV_PARAM_BUILDINFO, OSSL_PARAM_UTF8_PTR, &build, 0, (long unsigned int)&built_l },
     { NULL, 0, NULL, 0, (long unsigned int)NULL }
 };

 if ((prov = OSSL_PROVIDER_load(NULL, "oqsprovider")) != NULL
     && OSSL_PROVIDER_get_params(prov, request))
     printf("Provider 'oqsprovider' (%p) %s: %s\n", prov, OSSL_PROV_PARAM_BUILDINFO, build);
 else
     ERR_print_errors_fp(stderr);
}
```

outputs

```
Provider 'oqsprovider' (0x55fc2aa7db00) buildinfo: OQS Provider v.0.4.0-dev based on liboqs v.0.7.2-dev
```

**Reply by IProtex** [reply]
Hi @baentsch,

I tried the sample code, but it wasn't working on my side.

However, the printout from `ERR_print_errors_fp(stderr);` indicated that my program was trying to look up the `oqsprovider.so` from the _lib64/ossl-modules_ directory in my local OpenSSL 3 distribution (the one from the build).

**Placing the module library in this directory solved the problem and the loading of the provider now succeeds.**

Many thanks for your quick reply and help.

**Comment by baentsch** [comment, marked as answer]
Oh, and for the avoidance of doubt: You are sure that the library load itself succeeded? Did you set the environment variable `OPENSSL_MODULES` to point to the directory where you placed `oqsprovider.so`?

**Reply by IProtex** [reply]
Hi @baentsch,

It seems like that `dlopen` didn't properly loaded the module into my program :/

I haven't set the `OPENSSL_MODULES` environment variable, which had the consequence that OpenSSL was looking in the default OpenSSL module directory for the `oqsprovider.so`.

I placed the `oqsprovider.so` into the default module directory of my local OpenSSL distribution (_lib64/ossl-modules_) which solved the problem on my side :)

Thanks again for your quick reply and help.
